### PR TITLE
feat(crd-generator): Add support for more validation constraints

### LINKED ISSUE #5836: CRDGenerator: Add support for size constraints
### Is your enhancement related to a problem? Please describe

At the moment the CRDGenerator does not support size constraints:

- minItems
- minProperties
- minLength
- maxItems
- maxProperties
- maxLength

### Describe the solution you'd like

A new annotation `@Size` allows to define min and/or max constraints on a field or its accessors:

```java
@Target({ ElementType.ANNOTATION_TYPE, ElementType.FIELD, ElementType.METHOD })
@Retention(RetentionPolicy.RUNTIME)
public @interface Size {
  long min() default 0;
  long max() default Long.MAX_VALUE;
}
```

The CRDGenerator detects this annotation and adds JSONSchema constraints dependending on the java type:

| Java Type  | Constraints                    |
|------------|--------------------------------|
| String     | minLength<br>maxLength         |
| Array/List | minItems<br>maxItems           |
| Map        | minProperties<br>maxProperties |

If the annotation is used on an unsupported type, a warning message is logged.

### Describe alternatives you've considered

_No response_

### Additional context

Size constraints become more important if someone wants to use CEL Validation rules on fields of a CRD.
Because of the cost calculations of a CEL rule, constraints on strings/lists/maps are often only allowed if limits are defined.

### LINKED ISSUE #5868: CRDGenerator: Add support for exclusiveMinimum/exclusiveMaximum
### Is your enhancement related to a problem? Please describe

At the moment the CRDGenerator does not support exclusiveMinimum/exclusiveMaximum (boolean):

https://kubernetes.io/docs/reference/kubernetes-api/extend-resources/custom-resource-definition-v1/#JSONSchemaProps

**exclusiveMaximum**

[5.1.2.1.](https://json-schema.org/draft-04/draft-fge-json-schema-validation-00#rfc.section.5.1.2.1) Valid values
The value of "maximum" MUST be a JSON number. The value of "exclusiveMaximum" MUST be a boolean.

If "exclusiveMaximum" is present, "maximum" MUST also be present.

[5.1.2.2.](https://json-schema.org/draft-04/draft-fge-json-schema-validation-00#rfc.section.5.1.2.2) Conditions for successful validation
Successful validation depends on the presence and value of "exclusiveMaximum":

if "exclusiveMaximum" is not present, or has boolean value false, then the instance is valid if it is lower than, or equal to, the value of "maximum";
if "exclusiveMaximum" has boolean value true, the instance is valid if it is strictly lower than the value of "maximum".

[5.1.2.3.](https://json-schema.org/draft-04/draft-fge-json-schema-validation-00#rfc.section.5.1.2.3) Default value
"exclusiveMaximum", if absent, may be considered as being present with boolean value false.

**exclusiveMinimum**

[5.1.3.1.](https://json-schema.org/draft-04/draft-fge-json-schema-validation-00#rfc.section.5.1.3.1) Valid values
The value of "minimum" MUST be a JSON number. The value of "exclusiveMinimum" MUST be a boolean.

If "exclusiveMinimum" is present, "minimum" MUST also be present.

[5.1.3.2.](https://json-schema.org/draft-04/draft-fge-json-schema-validation-00#rfc.section.5.1.3.2) Conditions for successful validation
Successful validation depends on the presence and value of "exclusiveMinimum":

if "exclusiveMinimum" is not present, or has boolean value false, then the instance is valid if it is greater than, or equal to, the value of "minimum";
if "exclusiveMinimum" is present and has boolean value true, the instance is valid if it is strictly greater than the value of "minimum".

[5.1.3.3.](https://json-schema.org/draft-04/draft-fge-json-schema-validation-00#rfc.section.5.1.3.3) Default value
"exclusiveMinimum", if absent, may be considered as being present with boolean value false.

### Describe the solution you'd like

The existing annotations `@Min` and `@Max` allow to define the "inclusiveness":

```java
@Target({ ElementType.ANNOTATION_TYPE, ElementType.FIELD, ElementType.METHOD })
@Retention(RetentionPolicy.RUNTIME)
public @interface Min {
  /**
   * @return the element must be higher or equal to
   */
  double value();

  /**
   * Specifies whether the specified minimum is inclusive or exclusive.
   * By default, it is inclusive.
   *
   * @return {@code true} if the value must be higher or equal to the specified minimum,
   *         {@code false} if the value must be higher
   */
  boolean inclusive() default true;
}
```

```java
@Target({ ElementType.ANNOTATION_TYPE, ElementType.FIELD, ElementType.METHOD })
@Retention(RetentionPolicy.RUNTIME)
public @interface Max {
  /**
   * @return value the element must be lower or equal to
   */
  double value();

  /**
   *  Specifies whether the specified maximum is inclusive or exclusive.
   *  By default, it is inclusive.
   *
   * @return {@code true} if the value must be higher or equal to the specified minimum,
   *         {@code false} if the value must be higher
   */
  boolean inclusive() default true;
}
```

Note that this won't be a breaking change because the above defaults would result in `exclusiveMinimum` / `exclusiveMaximum` = `false` which can be omitted and which is the case at the moment.

### Describe alternatives you've considered

We could introduce instead or additionally a new annotation `@Schema` like in [swagger-core](https://github.com/swagger-api/swagger-core/wiki/Swagger-2.X---Annotations#schema), which allows to specify `exclusiveMinimum/exclusiveMaximum` directly.

But in my opinion it's more important to tie it directly to the value.

### Additional context

- https://kubernetes.io/docs/reference/kubernetes-api/extend-resources/custom-resource-definition-v1/#JSONSchemaProps
- https://json-schema.org/draft-04/draft-zyp-json-schema-04
- https://docs.swagger.io/swagger-core/v2.2.21/apidocs/io/swagger/v3/oas/annotations/media/Schema.html
- https://jakarta.ee/specifications/bean-validation/3.0/jakarta-bean-validation-spec-3.0.html#builtinconstraints-decimalmin

### PR DESCRIPTION
**Pull request by baloo42**
## Description

Fixes #5836 (Add support for size constraints)
Fixes #5868 (Add support for exclusiveMinimum/exclusiveMaximum)

Add support for the following validation constraints:

- exclusiveMinimum / exclusiveMaximum by extending `@Min` and `@Max` annotations
- minLength / maxLength for strings with new `@Size` annotation
- minItems / maxItems for array like containers with new `@Size` annotation
- minProperties / maxProperties for map like containers with `@Size` annotation

Notes:

- `@Size` will be detected on container types similar to `@Pattern` and `@Min/@Max`.
- The use of `@Min/@Max` and `@Size` is now restricted to their intended types,
  e.g. `@Min/@Max` is only intended to be used on number types and not on strings.
- This implementation uses regarding "exclusivness" the same strategy as described in [JSR-303](https://jakarta.ee/specifications/bean-validation/3.0/jakarta-bean-validation-spec-3.0.html#builtinconstraints-decimalmin):
  A boolean `inclusive` flag in the annotation, defaulting to `true`.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: crd-generator/test/src/test/java/io/fabric8/crd/generator/approvaltests/CRDGeneratorApprovalTest.java [hunk 1]
```diff
@@ -148,6 +152,14 @@ static Stream<TestCase> crdApprovalBaseCases(String crdVersion) {
     return cases.stream();
   }
 
+  static Stream<TestCase> crdApprovalCasesApiV2(String crdVersion) {
```
Comment (baloo42) [inline crd-generator/test/src/test/java/io/fabric8/crd/generator/approvaltests/CRDGeneratorApprovalTest.java:155]: @manusa I have started here with splitting test cases into two groups:  

- The existing tests for api-v1 and api-v2.
- Test cases targeting new features in api-v2 only.


I think this change is also important for other feature additions. Please let me know if I should prepare this in an extra PR to be able to merge this faster.

Comment (baloo42) [reply crd-generator/test/src/test/java/io/fabric8/crd/generator/approvaltests/CRDGeneratorApprovalTest.java:157]: see https://github.com/fabric8io/kubernetes-client/pull/6533

**Review by metacosm — APPROVED** [review]
Looks reasonable to me.

**Comment by baloo42** [comment]
@andreaTP can you have a look on this?

You said [here](https://github.com/fabric8io/kubernetes-client/issues/5868#issuecomment-2047117058
):

> I would use exclusive as opposed to inclusive as it better matches JSONSchema terminology.

Are you fine with this implementation (which follows common annotations) or should we follow the JSONSchema terminology?

@manusa Do you have an opinion on this?

Other open discussion topics are:

- Should `@Size` be detected on container types, if the type itself is a container?
  e.g. `List<@Size(min=1) List<@Size(min=2) String>>`
- I think our `@Min`/`@Max` annotation has the wrong type for the value. It feels wrong that we are using `double`. Are there any plans or thoughts to tackle this? Should we create a new issue for this? (would be a breaking change)

**Comment by baloo42** [comment]
> > @manusa Do you have an opinion on this?
> I really don't care as long as it's intuitive for users.

We are following Jakarta Bean Validation here, so this PR should be fine in terms of your goal.

> Since this doesn't involve any breaking changes I think it's fine if we delay it to 7.1.

If the following change is considered as bugfix, then we can postpone it:

> The use of `@Min/@Max` and `@Size` is now restricted to their intended types,
>  e.g. `@Min/@Max` is only intended to be used on number types and not on strings.

---
> What I'd really like is to release v7.0 this week. I'm not sure if there's anything else left.

I have not found anything else breaking, so feel free to create the release with or without this PR.

**Comment by manusa** [comment]
> If the following change is considered as bugfix, then we can postpone it:

OK; I forgot about that change. I'm merging it now.

Maybe you want to add something in the MIGRATIOM-v7.md section for the CRD Generator.

**Pull request merged**